### PR TITLE
use threadgroup pointers instead of references in metal

### LINKED ISSUE #4500: [msl-out] Incorrect behavior over workgroup memory due to possibly miscompiled barrier in Metal
I believe I have identified a bug in the Metal compiler that violates the expected memory access and barrier ordering when a compute kernel uses dynamic threadgroup memory (i.e. when threadgroup shared memory is declared as an entry-point parameter rather than a fixed-size function-local variable). I have a very reliable reproduction (in Metal) of this bug implemented and detailed at https://github.com/armansito/metal-workgroup-memory-bug. A reproduction of the bug
using wgpu can be found here: https://github.com/armansito/naga-workgroup-memory-bug.

I ran into this while debugging a system-wide freeze due to one of the shaders in linebender/vello that uses the result of `workgroupUniformLoad` from shared memory to exit a compute stage. The miscompilation causes the load to be non-uniform which violates an important invariant. The bug is reproducible using wgpu but not when the same shaders are executed by Dawn. I believe this is because Naga implements workgroup buffers as dynamic (i.e. an entry-point parameter) while Tint does not. I think this is also possibly related to https://github.com/gfx-rs/wgpu/issues/3181.

A WGSL version of the [Metal program from my test case](https://github.com/armansito/metal-workgroup-memory-bug) looks like this:

```WGSL
@group(0) @binding(0) var<uniform> flag: u32;
@group(0) @binding(1) var<storage, read_write> output: atomic<u32>;

var<workgroup> shared_flag: u32;

@compute @workgroup_size(64)
fn main(@builtin(local_invocation_id) local_id: vec3<u32>) {
    if local_id.x == 0u {
        shared_flag = 0xffffffffu;
    }
    workgroupBarrier();

    if local_id.x == 0u {
        shared_flag = flag;
    }

    let abort = workgroupUniformLoad(&shared_flag);
    if abort != 0u {
        return;
    }

    atomicAdd(&output, 1u);
}
```

The test case sets up `flag` to always have a value of 0. The expected behavior is for all threads in the workgroup to add 1 to `output` (which is what I see with Dawn/Tint). The observed behavior is that seemingly only one SIMD group increments the atomic in all workgroups.

This is obviously not a bug in Naga but in the Metal driver but since it impacts correctness I propose that Naga take the same approach as Tint and convert workgroup variables to a function-local threadgroup variable instead of an entry-point parameter.

### PR DESCRIPTION
**Pull request by 39ali**
**Connections**
fixes #4500 

**Description**
threadgroups are broken in metal when referenced instead of using pointers(i suspect because of compiler reordering) 
this seems to fix the issue 

**Testing**
ran tests, and local test that uses threadgroups

**Squash or Rebase?**
either

**Checklist**

- [x] Run `cargo fmt`.
- [ ] Run `taplo format`.
- [ ] Run `cargo clippy --tests`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [ ] If this contains user-facing changes, add a `CHANGELOG.md` entry.

### CODE REVIEW

**Comment by ErichDonGubler** [comment]
@39ali: Before we have somebody do review, this needs to be rebased, and CI errors need to be fixed.

File: naga/src/back/msl/writer.rs [hunk 1]
```diff
@@ -454,14 +454,21 @@ impl TypedGlobalVariable<'_> {
                 } else {
                     ""
                 };
-                (coherent, space, access, "&")
+
+                let (suffix, reference) = if let crate::AddressSpace::WorkGroup = var.space {
```
Comment (inner-daemons) [inline naga/src/back/msl/writer.rs]: If let is probably less clear than simply `var.space == ...`. Also, do you know whether this should affect `AddressSpace::TaskPayload`, or any ray payload space? If not that warrants a comment explaining.

File: naga/src/back/msl/writer.rs [hunk 2]
```diff
@@ -7607,6 +7603,27 @@ template <typename A>
             // end of the entry point argument list
             writeln!(self.out, ") {{")?;
 
+            // for threadgroup, we use pointer and not a reference to disable compiler reordering
+            // but for it to work with the rest of the code we reference it in a temp var in the function body:
+            // threadgroup type& temp = *temp_ptr;
+            for (handle, var) in module.global_variables.iter() {
+                if var.space == crate::AddressSpace::WorkGroup && !fun_info[handle].is_empty() {
```
Comment (inner-daemons) [inline naga/src/back/msl/writer.rs]: Again, TaskPayload?

File: naga/src/back/msl/writer.rs [hunk 3]
```diff
@@ -7607,6 +7603,27 @@ template <typename A>
             // end of the entry point argument list
             writeln!(self.out, ") {{")?;
 
+            // for threadgroup, we use pointer and not a reference to disable compiler reordering
+            // but for it to work with the rest of the code we reference it in a temp var in the function body:
+            // threadgroup type& temp = *temp_ptr;
```
Comment (inner-daemons) [inline naga/src/back/msl/writer.rs]: This should probably link to an issue. "Compiler reordering" is not perfectly clear on what the problem actually is. And its not obvious why we can redeclare as a refernece later without issue.

File: naga/src/back/msl/writer.rs [hunk 4]
```diff
@@ -454,14 +454,21 @@ impl TypedGlobalVariable<'_> {
...
+                (coherent, space, access, suffix, reference)
             }
-            _ => ("", "", "", ""),
+            _ => ("", "", "", "", ""),
         };
 
         Ok(write!(
             out,
-            "{}{}{}{}{}{}{} {}",
+            "{}{}{}{}{}{}{} {}{}",
```
Comment (inner-daemons) [inline naga/src/back/msl/writer.rs]: This does not adequately handle cases where there might be another variable with this name. You have to use the writer's namer.

File: naga/src/back/msl/writer.rs [hunk 5]
```diff
@@ -7607,6 +7603,27 @@ template <typename A>
...
+                        handle: var.ty,
+                        gctx: module.to_ctx(),
+                        names: &self.names,
+                        access: crate::StorageAccess::default(),
+                        first_time: false,
+                    };
+
+                    writeln!(
+                        self.out,
+                        "    threadgroup {}& {} = *{}_ptr;",
```
Comment (inner-daemons) [inline naga/src/back/msl/writer.rs]: This _ptr will have to change according to the comment above.

**Review by inner-daemons — CHANGES_REQUESTED** [review]
Some initial thoughts, nothing major but will need to be addressed

**Comment by 39ali** [comment]
@inner-daemons  i didn't like that local reference hack so i changed it to actually use the pointer

File: naga/src/back/msl/writer.rs (hunk 1 above)
Comment (39ali) [reply naga/src/back/msl/writer.rs]: just saw the mesh shader changes, i'll look into it

**Comment by teoxoy** [comment]
@39ali could you resolve the CI failures? I will take a look at the PR after.

**Comment by 39ali** [comment]
@teoxoy done

**Comment by inner-daemons** [comment]
Gonna block this on my review again since it touches mesh shader stuff

File: naga/src/back/msl/writer.rs [hunk 6]
```diff
@@ -3143,6 +3143,31 @@ impl<W: Write> Writer<W> {
...
+                let arg = &context.function.arguments[index as usize];
+                let type_inner = &context.module.types[arg.ty].inner;
+                matches!(
+                    type_inner,
+                    crate::TypeInner::Pointer {
+                        space: crate::AddressSpace::WorkGroup,
+                        ..
+                    }
+                )
+            }
```
Comment (teoxoy) [inline naga/src/back/msl/writer.rs]: The validator only allows arguments in the private and function address spaces.

File: naga/src/back/msl/writer.rs [hunk 7]
```diff
@@ -8043,19 +8101,54 @@ mod workgroup_mem_init {
...
+                                module.types[member.ty].inner,
+                                crate::TypeInner::Pointer { .. }
+                            )
+                        } else {
+                            false
+                        }
+                    } else {
+                        false
+                    }
+                }
```
Comment (teoxoy) [inline naga/src/back/msl/writer.rs]: Structs can't have pointers in WGSL, those are validated out.

**Review by teoxoy — COMMENTED** [review]
Looks good overall, but I think we can remove 2 sections.

File: naga/src/back/msl/writer.rs (hunk 7 above)
Comment (39ali) [reply naga/src/back/msl/writer.rs]: oh right , i'll fix it

**Comment by 39ali** [comment]
@teoxoy  done 👍

**Comment by teoxoy** [comment]
Thanks!

**Review by teoxoy — APPROVED** [review]

File: naga/tests/out/msl/wgsl-mesh-shader.metal [hunk 8]
```diff
@@ -74,9 +74,9 @@ metal::uint3 _ts_main(
   metal::mesh_grid_properties nagaMeshGrid
 , uint __local_invocation_index [[thread_index_in_threadgroup]]
 , object_data TaskPayload& taskPayload [[payload]]
-, threadgroup float& workgroupData
 ) {
-    uint3 nagaGridSize = _ts_main(__local_invocation_index, taskPayload, workgroupData);
+    threadgroup float workgroupData;
```
Comment (inner-daemons) [inline naga/tests/out/msl/wgsl-mesh-shader.metal:78]: Why did you make this change

File: naga/tests/out/msl/wgsl-mesh-shader.metal [hunk 9]
```diff
@@ -54,15 +54,15 @@ void helper_writer(
 metal::uint3 _ts_main(
   uint __local_invocation_index
 , object_data TaskPayload& taskPayload
-, threadgroup float& workgroupData
+, threadgroup float* workgroupData
```
Comment (inner-daemons) [inline naga/tests/out/msl/wgsl-mesh-shader.metal:57]: Is this really necessary? Keep in mind this right here isn't actually declaring such a workgroup variable, its merely taking a reference to an existing variable

File: naga/tests/out/msl/wgsl-mesh-shader.metal (hunk 8 above)
Comment (inner-daemons) [reply naga/tests/out/msl/wgsl-mesh-shader.metal:78]: I'll give you some more context. There is a difference between workgroup variables declared as function parameters and those declared as variables in the function, because you have to manually allocate those declared in the function parameters (and then bind them). But there are benefits so we usually use those.

Mesh shaders don't support using this kind of workgroup variables because the grids can be dispatched by another task shader (which obviously can't allocate and bind a workgroup variable for each dispatched mesh shader grid). So mesh shaders intentionally declare them as variables in the wrapper function, but the inner function doesn't notice since it just takes a reference anyway. Task shaders don't have this limitation so they are treated like compute shaders.

File: naga/src/back/msl/writer.rs [hunk 10]
```diff
@@ -442,7 +442,7 @@ impl TypedGlobalVariable<'_> {
         };
         let (coherent, space, access, reference) = match (var.space.to_msl_name(), var.space) {
             (Some(space), crate::AddressSpace::WorkGroup) => {
```
Comment (inner-daemons) [inline naga/src/back/msl/writer.rs:444]: Should this also match for TaskPayload address space?

File: naga/src/back/msl/writer.rs [hunk 11]
```diff
@@ -3143,6 +3143,20 @@ impl<W: Write> Writer<W> {
...
 
+    fn is_root_workgroup_pointer(
+        &self,
+        chain: Handle<crate::Expression>,
+        context: &ExpressionContext,
+    ) -> bool {
+        match context.function.expressions[chain] {
+            crate::Expression::GlobalVariable(handle) => {
+                let var = &context.module.global_variables[handle];
+                var.space == crate::AddressSpace::WorkGroup
```
Comment (inner-daemons) [inline naga/src/back/msl/writer.rs:3154]: Same here, should this match for task payload variables?

File: naga/src/back/msl/writer.rs [hunk 12]
```diff
@@ -7449,7 +7495,8 @@ template <typename A>
                     }
                     _ => {
                         if var.space == crate::AddressSpace::WorkGroup
-                            && ep.stage == crate::ShaderStage::Mesh
+                            && (ep.stage == crate::ShaderStage::Mesh
```
Comment (inner-daemons) [inline naga/src/back/msl/writer.rs:7498]: This is the kind of stuff that I highlighted in the mesh shader ouptut but I don't know why it needed to change

File: naga/src/back/msl/writer.rs [hunk 13]
```diff
@@ -8043,19 +8090,35 @@ mod workgroup_mem_init {
     }
 
     impl Access {
+        fn is_pointer_type(&self, module: &crate::Module) -> bool {
+            match *self {
+                Access::GlobalVariable(handle) => {
+                    let var = &module.global_variables[handle];
+                    var.space == crate::AddressSpace::WorkGroup
```
Comment (inner-daemons) [inline naga/src/back/msl/writer.rs:8097]: Again here, should this match a task payload pointer

File: naga/src/back/msl/writer.rs [hunk 14]
```diff
@@ -8094,12 +8157,33 @@ mod workgroup_mem_init {
...
+                };
+                next.write(writer, names, op)?;
             }
             Ok(())
         }
+
+        fn root_is_workgroup_pointer(&self, module: &crate::Module) -> bool {
+            if let Some(&Access::GlobalVariable(handle)) = self.stack.first() {
+                let var = &module.global_variables[handle];
+                return var.space == crate::AddressSpace::WorkGroup;
```
Comment (inner-daemons) [inline naga/src/back/msl/writer.rs:8183]: You know the deal

**Review by inner-daemons — CHANGES_REQUESTED** [review]
I left some comments above, didn't realize there would be so many so I didn't put it in a review. Here, I mostly want you to avoid making changes to mesh/task shaders that aren't necessary.

Another thing that came up a ton is that `TaskPayload` address space functions very similarly to workgroup address space (though it is only mutable in task shaders). So it would probably be best to apply all of your fixes to that too.

File: naga/tests/out/msl/wgsl-mesh-shader.metal (hunk 8 above)
Comment (teoxoy) [reply naga/tests/out/msl/wgsl-mesh-shader.metal:78]: Good catch! I wonder why none of the tests failed.

Comment (inner-daemons) [reply naga/tests/out/msl/wgsl-mesh-shader.metal:78]: @teoxoy There was an edge case here to protect mesh shaders. It was expanded to task shaders in this change, not removed entirely. So it shouldn't have broken anything but it might've worsened performance or reduced the maximum size of the payload unexpectedly, I'm not entirely sure.

File: naga/src/back/msl/writer.rs (hunk 10 above)
Comment (teoxoy) [reply naga/src/back/msl/writer.rs:444]: I don't think it should, it doesn't map to the threadgroup address space.

https://github.com/gfx-rs/wgpu/blob/253477b9ec633d4a135821d139eb4dd527110b2b/naga/src/back/msl/writer.rs#L688

Comment (inner-daemons) [reply naga/src/back/msl/writer.rs:444]: Task payload storage class is defined to be basically identical to threadgroup storage class, except that in mesh shaders it is immutable. My point here is that this fix should therefore also apply to object_data variables so that they can benefit.

**Comment by inner-daemons** [comment]
This is currently waiting on author

**Comment by teoxoy** [comment]
@39ali will you be able to address @inner-daemons's comments?

**Comment by teoxoy** [comment]
If not maybe we can open a separate issue to do this for task and mesh shaders but at least https://github.com/gfx-rs/wgpu/pull/9380#discussion_r3128609110 should be resolved before landing.

**Comment by inner-daemons** [comment]
I would push very hard for this pull request to address mesh shaders since I worry about them always being put second and allowed to get worse, and it shouldn't be too hard to fix in this case.

**Comment by inner-daemons** [comment]
@39ali Do you plan to come back to this PR?
